concordium-std-derive/src/lib.rs

```diff
@@ -1939,6 +1943,113 @@ fn concordium_test_worker(_attr: TokenStream, item: TokenStream) -> syn::Result<
     Ok(test_fn.into())
 }
 
+/// Look up the `tests` identifier in `NestedMeta` and return the value
+/// associated with it. If no `tests` is found or parsing the value has failed,
+/// return a error
+#[cfg(feature = "concordium-quickcheck")]
+fn get_quickcheck_tests_count(meta: &NestedMeta) -> Result<u64, syn::parse::Error> {
+    match meta {
+        NestedMeta::Meta(Meta::NameValue(v)) => {
+            if v.path.is_ident("tests") {
+                match &v.lit {
+                    Lit::Int(i) => i
+                        .base10_parse::<u64>()
+                        .map_err(|_| syn::parse::Error::new(i.span(), "error when pasing integer")),
+                    _ => Err(syn::parse::Error::new(v.span(), "unexpected attribute")),
+                }
+            } else {
+                Err(syn::parse::Error::new(v.span(), "unexpected attribute"))
+            }
+        }
+        NestedMeta::Meta(m) => Err(syn::parse::Error::new(m.span(), "unexpected attribute")),
+        NestedMeta::Lit(l) => {
+            Err(syn::parse::Error::new(l.span(), "expected a named attribute, e.g. tests = 1000"))
+        }
+    }
+}
+
+/// A default number of tests to run with QuickCheck
+#[cfg(feature = "concordium-quickcheck")]
+const DEFAULT_QUICKCHECK_TESTS: u64 = 100;
+
+#[cfg(feature = "concordium-quickcheck")]
+#[proc_macro_attribute]
+/// Derive the appropriate export for an annotated QuickCheck function by
+/// exposing it as `#[concordium_test]`. The macro was copied from `QuickCheck`
+/// and changed to use `concordium_std::test_infrastructure::concordium_qc`
+/// instead of the standard  `QuickCheck`'s `quickcheck`
+///
+/// The macro takes a `tests` attrubute that specifies how many tests to run:
+/// `#[concordium_quickcheck(tests = 1000)]`. If no `tests` is provided,
+/// `DEFAULT_QUICKCHECK_TESTS` is used
+pub fn concordium_quickcheck(attr: TokenStream, input: TokenStream) -> TokenStream {
+    use syn::AttributeArgs;
+
+    let output = match syn::Item::parse.parse(input.clone()) {
+        Ok(syn::Item::Fn(mut item_fn)) => {
+            // Parse attrubutes and fail early if wrong atributes are specified
+            let a = attr.clone();
+            let parsed_attr = parse_macro_input!(a as AttributeArgs);
+            let tests = if parsed_attr.is_empty() {
+                DEFAULT_QUICKCHECK_TESTS
+            } else if parsed_attr.len() == 1 {
+                let meta = parsed_attr.get(0).unwrap();
+                match get_quickcheck_tests_count(meta).map_err(|e| e.to_compile_error()) {
+                    Ok(v) => v,
+                    Err(e) => return e.into(),
+                }
+            } else {
+                return syn::parse::Error::new(
+                    proc_macro2::TokenStream::from(attr).span(),
+                    "expected a single `tests = <number>` attribute",
+                )
+                .to_compile_error()
```

Comment:
It is a bit noisy with all the errors, converting to compile error and making sure to return early.
I think the code would be more readable if you refactor the logic out into a function returning syn::Result, so you can use `?`. Then in the top level function you would trigger the logic one, convert errors to compile error and so on.
Let me know if this does not make sense

Reply:
@limemloh I've factored out some logic, I think it looks better now, but moving out more logic doesn't help much.

Another option is something like this:

```Rust
#[cfg(feature = "concordium-quickcheck")]
fn wrap_quickcheck_test(
    attr: TokenStream,
    item_fn: &mut syn::ItemFn,
) -> syn::Result<proc_macro2::TokenStream> {
    let num_tests: u64 = parse_quickcheck_num_tests(attr)?;
    let mut inputs: Punctuated<syn::BareFnArg, syn::token::Comma> =
        syn::punctuated::Punctuated::new();

    for input in item_fn.sig.inputs.iter() {
        match input {
            syn::FnArg::Typed(syn::PatType {
                ref ty,
                ..
            }) => {
                inputs.push(parse_quote!(_: #ty));
            }
            syn::FnArg::Receiver(_) => {
                return Err(syn::parse::Error::new(
                    input.span(),
                    "`self` arguments are not supported",
                ))
            }
        }
    }

    let attrs = mem::take(&mut item_fn.attrs);
    let name = &item_fn.sig.ident;
    let codomain = &item_fn.sig.output;
    let res = quote! {
        #[concordium_test]
        #(#attrs)*
        fn #name() {
            #item_fn
           ::concordium_std::test_infrastructure::concordium_qc(#num_tests, #name as (fn (#inputs) #codomain))
        }
    };
    Ok(res)
}

#[cfg(feature = "concordium-quickcheck")]
#[proc_macro_attribute]
pub fn concordium_quickcheck(attr: TokenStream, input: TokenStream) -> TokenStream {
    let output = match syn::Item::parse.parse(input.clone()) {
        Ok(syn::Item::Fn(mut item_fn)) => {
            wrap_quickcheck_test(attr, &mut item_fn).unwrap_or_else(|e| e.to_compile_error())
        }
        Ok(_) => {
            let span = proc_macro2::TokenStream::from(input).span();
            let msg = "#[concordium_quickcheck] can only be applied to functions.";

            syn::parse::Error::new(span, msg).to_compile_error()
        }
        Err(e) => e.to_compile_error(),
    };

    output.into()
}

```

Reply:
I think your suggestion here is a lot more readable

Reply:
Cool, I've improved it a bit, I'll push the changes then.